From 5c27eb8309b371907eecd4f5666d513f159c6842 Mon Sep 17 00:00:00 2001
From: Peng Tao <tao.peng@emc.com>
Date: Tue, 26 Feb 2013 18:09:42 +0800
Subject: [PATCH 5/6] LU-2850 kernel: 3.8 upstream removes vmtruncate()

vmtruncate() is removed since upstream commit b9f61c3.
We can open code is for llite though.

Signed-off-by: Peng Tao <tao.peng@emc.com>
Change-Id: I9d4f43a5c181f16482b4caa021ba7a09ee52b600
---
 lustre/autoconf/lustre-core.m4 | 23 +++++++++++++++++++++++
 lustre/llite/vvp_io.c          | 10 +++++++++-
 2 files changed, 32 insertions(+), 1 deletion(-)

diff --git a/lustre/autoconf/lustre-core.m4 b/lustre/autoconf/lustre-core.m4
index deec2d7..94e57f0 100644
--- a/lustre/autoconf/lustre-core.m4
+++ b/lustre/autoconf/lustre-core.m4
@@ -2278,6 +2278,26 @@ EXTRA_KCFLAGS="$tmp_flags"
 ])
 
 #
+# 3.8 removes vmtruncate()
+# see upstream commit b9f61c3
+#
+AC_DEFUN([LC_HAVE_MM_VMTRUNCATE],
+[AC_MSG_CHECKING([if mm has vmtruncate])
+LB_LINUX_TRY_COMPILE([
+	#include <linux/mm.h>
+],[
+	vmtruncate(NULL, 0);
+],[
+	AC_DEFINE(HAVE_MM_VMTRUNCATE, 1,
+		[vmtruncate is defined by the kernel])
+	AC_MSG_RESULT([yes])
+],[
+	AC_MSG_RESULT([no])
+])
+])
+
+
+#
 # LC_PROG_LINUX
 #
 # Lustre linux kernel checks
@@ -2462,6 +2482,9 @@ AC_DEFUN([LC_PROG_LINUX],
 	 LC_HAVE_SOCK_ALLOC_FILE
 	 LC_HAVE_STRUCT_FILENAME
 
+	 # 3.8
+	 LC_HAVE_MM_VMTRUNCATE
+
 	 #
 	 if test x$enable_server = xyes ; then
 		AC_DEFINE(HAVE_SERVER_SUPPORT, 1, [support server])
diff --git a/lustre/llite/vvp_io.c b/lustre/llite/vvp_io.c
index 9020d77..25f631c 100644
--- a/lustre/llite/vvp_io.c
+++ b/lustre/llite/vvp_io.c
@@ -362,9 +362,17 @@ static int vvp_do_vmtruncate(struct inode *inode, size_t size)
 	 * Only ll_inode_size_lock is taken at this level.
 	 */
 	ll_inode_size_lock(inode);
+#ifdef HAVE_MM_VMTRUNCATE
 	result = vmtruncate(inode, size);
+#else
+	result = inode_newsize_ok(inode, size);
+	if (result < 0) {
+		ll_inode_size_unlock(inode);
+		return result;
+	}
+	truncate_setsize(inode, size);
+#endif
 	ll_inode_size_unlock(inode);
-
 	return result;
 }
 
-- 
1.8.1.5

